### DISCUSSION TITLE
Filter "body" should have access to "system message"

### DISCUSSION THREAD

**Discussion by clang88** [open-webui/open-webui#6930] [General]
# Feature Request
I am trying to build a "token counter" filter for Open WebUI; it works fine, adding the "Tokens" at the end of each System output but removing them from the messages again before sending the message to the LLM.

The issue is: The token count is incorrect, because the "body" object does not seem to contain the system message:
![image](https://github.com/user-attachments/assets/29ef64e0-e887-4d42-be17-df083c220ea5)

---

**Is your feature request related to a problem? Please describe.**
This is particularly important when using RAG, as the context is added into the system message and is therefore not counted by my "token counter" filter.

**Describe the solution you'd like**
I believe it would be beneficial to also have acces to the system message with the filter, i.e. add the system message to the chain of messages. 

**Describe alternatives you've considered**
Perhaps we should just add a "Token counter" for openAI models as is already present for Ollama models?

**Comment by clang88** [comment]
This is the code for the filter, btw. I commented out the trick I used to see the body contents.
```python
"""
title: Token Counter
author: clang88
version: 1.0
"""

from pydantic import BaseModel, Field
from typing import Optional
import tiktoken
import json
import os



class Filter:
    class Valves(BaseModel):
        priority: int = Field(
            default=0, description="Priority level for the filter operations."
        )
        pass

    def __init__(self):
        self.valves = self.Valves()
        self.input_cost = 0
        self.output_cost = 0
        pass

    def inlet(self, body: dict, __user__: Optional[dict] = None) -> dict:
        if __user__.get("role", "admin") in ["user", "admin", "system"]:
            messages = body.get("messages", [])
            model = body.get("model")
            try:
                enc = tiktoken.encoding_for_model(model)
            except Exception as e:
                if "gpt-3.5-turbo" in model:
                    enc = tiktoken.encoding_for_model("gpt-3.5-turbo-")
                else:
                    print(f"Error: {e}")
                    enc = tiktoken.encoding_for_model("gpt-4o")

            for message in messages:
                if "\n\n**Token consumption:**" in message["content"]:
                    message["content"] = "".join(
                        message["content"].split("\n\n**Token consumption:**")
                    )[:-1]
        return body

    def outlet(self, body: dict, __user__: Optional[dict] = None) -> dict:
        messages = body.get("messages", [])
        model = body.get("model")
        try:
            enc = tiktoken.encoding_for_model(model)
        except Exception as e:
            if "gpt-3.5-turbo" in model:
                enc = tiktoken.encoding_for_model("gpt-3.5-turbo-")
            else:
                print(f"Error: {e}")
                enc = tiktoken.encoding_for_model("gpt-4o")

        total_tokens = sum(len(enc.encode(message["content"])) for message in messages)
        # messages[-1]["content"] = str(body)
        messages[-1]["content"] = str(
            messages[-1]["content"] + "\n\n**Token consumption:** " + str(total_tokens)
        )

        return body
```

**Comment by clang88** [comment]
Am I really the only one interested in this? I think access to the system prompt when using filters is essential to be able to do certain workflows. E.g. change the system prompt based on the request of the user. I found some snippets on openwebui.com, where people would simply inject a system prompt as the first message in `body["messages"]`, but that seems kind of strange... Are two system prompts passed to the API in that case?

**Comment by clang88** [comment]
Still really surprised no one is replying here, given the system message is such a major tool to steer the way the LLM works. I played around a bit with the filters function and mostly reverse engineered what is going on:
So it appears that the "body" passed to the "inlet" function will have a "system" message, if the system message is added via the advanced controls. If the model has the system message added in the model configuration, it is found in `body["metadata"]["model"].params.system` (the "metadata" model is a "ModelModel" object). So far so good, but unfortunately even if I change the system prompt via `body["metadata"]["model"].params.system = "Always answer with "Duh!""` the original system prompt (Always answer with "Ok!") seems to be transferred to the API.

I mean, I can live with it, if custom models have their system prompt protected from being tampered with by the filters, just need to know it. In that case I could create a model which simply has no system prompt and add it as a message depending on the task the model should perform.

Use-Case: Assume I want to create a model that can do "Translation", "Transliteration" and "Transcreation"; for these 3 the system prompt would look quite differently, so it would be great if, via filter, I could set the system prompt, depending on what the user asked.

Edit: It appears that the messages are created differently for ollama and openai. I have this system prompt in the test model: "Answer with OK!" and the following in the filter function "and also say DUH!".

Both models seem to say "OK! DUH!" when prompted, but here's the strange thing. 
This is the inlet messages for Ollama: messages: [{'role': 'system', 'content': 'and also say DUH!'}, {'role': 'user', 'content': 'test with llama 3.2'}]
This is the inlet meesages for OpenAI: messages: [{'role': 'system', 'content': 'Answer with "OK!"\nand also say DUH!'}, {'role': 'user', 'content': 'test with gpt4o-mini'}]

Notice how the model system prompt is only shown in the OpenAI messages. However, it is clear that the model system prompt is reaching ollama as well, judging by the answer of the model:
Ollama: {'id': '3af45db1-d548-43f1-9a66-a88dad006c68', 'role': 'assistant', 'content': "**OK!\n\nDUH!** It looks like you're testing the version of my language model. Llama 3.2 is a relatively recent update, but I'm not sure what specific changes it brings. Can you tell me more about what you'd like to know or discuss?", 'timestamp': 1739579799}
OpenAI: {'id': 'e32df198-9b95-4b74-b019-a26fd9f91052', 'role': 'assistant', 'content': '**OK! DUH!**', 'timestamp': 1739579817}]

**Reply by kaptainkangaroo** [reply]
I'm having similar struggles, trying to update the system prompt in the "chat" table record based on the chat_id. there are no messages with type "system" only "user" and "assistant"  I don't understand where the system message is stored and how its resent to openai for each api call.  in other words, I can use the filter to append a string to the system prompt, but I need to do it on every api call, which is highly inefficient - because I want to just do it once upon the  thread initialization

it appears the system message is never stored in the chat or message table, instead its dynamically generated and inserted for each api call.

**Comment by kaytwo** [comment]
I was able to get my changes to the system prompt during an inlet filter to persist with this patch:

```patch
diff --git a/backend/open_webui/functions.py b/backend/open_webui/functions.py
index 2f94f701e..2ce2e0369 100644
--- a/backend/open_webui/functions.py
+++ b/backend/open_webui/functions.py
@@ -251,7 +251,10 @@ async def generate_function_chat_completion(
         if model_info.base_model_id:
             form_data["model"] = model_info.base_model_id

-        params = model_info.params.model_dump()
+        params = {
+            **model_info.params.model_dump(),
+            **metadata["model"]["info"]["params"],
+        }
         form_data = apply_model_params_to_body_openai(params, form_data)
         form_data = apply_model_system_prompt_to_body(params, form_data, metadata, user)

diff --git a/backend/open_webui/utils/chat.py b/backend/open_webui/utils/chat.py
index 1a9d14248..ed1099ba4 100644
--- a/backend/open_webui/utils/chat.py
+++ b/backend/open_webui/utils/chat.py
@@ -170,8 +170,8 @@ async def generate_chat_completion(
             form_data["metadata"] = request.state.metadata
         else:
             form_data["metadata"] = {
-                **form_data["metadata"],
                 **request.state.metadata,
+                **form_data["metadata"],
             }

     if getattr(request.state, "direct", False) and hasattr(request.state, "model"):
diff --git a/backend/open_webui/utils/middleware.py b/backend/open_webui/utils/middleware.py
index f375a39f0..080daa748 100644
--- a/backend/open_webui/utils/middleware.py
+++ b/backend/open_webui/utils/middleware.py
@@ -760,7 +760,7 @@ async def process_chat_payload(request, form_data, user, metadata, model):
         files = list({json.dumps(f, sort_keys=True): f for f in files}.values())

     metadata = {
-        **metadata,
+        **form_data.get("metadata", metadata),
         "tool_ids": tool_ids,
         "files": files,
     }
```

Following the execution of filters (and in my case, pipes, as I'm using a Gemini model), it appears that the metadata that gets returned in the form body from the filter keeps getting overwritten in multiple places along the way to actually calling the base model. I'm not sure if these changes will break anything else, but they were able to allow me to see my edited prompt persist through to the pipe model call.

I doubt it would be worth it to submit a pull request with these changes as I really don't understand the underlying code, but if you want to do some further testing and submit it please do.

**Comment by DmitriyAlergant** [comment]
I have encountered the same limitation, and I do think it needs a Pull Request. @clang88 can you still submit yours? Even if @tjbck disagrees with some of the implementation intricacies, the PR will bring his attention to the issue.

What's interesting is that Open WebUI tends to embed attached file contents also as part of the system message - at least in recent versions (0.6.5). You can add logging of body[] inside the filter inlet and see for yourself... The attached file context (no-RAG) is added to the system message wrapped into RAG Template prompt, and embedded file content. It all currently goes into the system message for some reason. 

And the token counting (or moderation, guardrails) inlet filter basically can't access the file content as sent to the LLM. Unless it relies on file ID metadata and separately retrieves the files, etc; But that's still not the same body content that the LLM will see.

**Comment by rvkwi** [comment]
Did anything change about this since?

Just ran into this while trying to get to the system prompt from a filter. 
`__messages__` like in tools does not exist, and the `body["messages"]` unlike `__messages__` does not contain the system prompt. Couldn't find it in the `metadata` either for a filter.  I used to be able to get it through `body["metadata"]["model"]["info"]["params"]["system"]` at some point, but seems in the current one that doesn't work either anymore. 

Any way to do this in a filter like you can in tools?

**Reply by rvkwi** [reply]
Found it, nevermind, it's accessible by resolving it through the router in a filter.

```
"""
title: Get System Prompt Filter
author: open-webui
author_url: https://github.com/open-webui
funding_url: https://github.com/open-webui
version: 0.1
"""

from pydantic import BaseModel, Field
from typing import Optional
from open_webui.models.models import Models


class Filter:
    class Valves(BaseModel):
        pass

    def __init__(self):
        self.valves = self.Valves()

    async def inlet(
        self,
        body: dict,
        __user__: Optional[dict] = None,
        __event_emitter__=None,
        __metadata__: Optional[dict] = None,
        __model__: Optional[dict] = None,
    ) -> dict:

        model_id = body.get("model")
        model = Models.get_model_by_id(model_id)
        system_prompt = model.params.system if model and model.params else None
        print(system_prompt)

        return body

```

if anyone needs it.
